core/src/main/java/io/questdb/griffin/engine/functions/eq/EqVarcharFunctionFactory.java

```diff
@@ -118,24 +118,49 @@ public void toPlan(PlanSink sink) {
         }
     }
 
-    private static class Func extends AbstractEqBinaryFunction {
-        public Func(Function left, Function right) {
+    static class Func extends AbstractEqBinaryFunction {
+        Func(Function left, Function right) {
             super(left, right);
         }
 
         @Override
         public boolean getBool(Record rec) {
-            // important to compare A and B varchars in case
-            // these are columns of the same record
-            // records have re-usable character sequences
+            // getVarcharA/B returns a reusable sequence object.
+            // When using two at once, it's important to use both A and B.
             final Utf8Sequence a = left.getVarcharA(rec);
             final Utf8Sequence b = right.getVarcharB(rec);
-
             if (a == null) {
                 return negated != (b == null);
             }
+            return negated != Utf8s.equals(a, b);
+        }
+
+        @Override
+        public String getName() {
+            if (negated) {
+                return "!=";
+            } else {
+                return "=";
+            }
+        }
+    }
+
+    static class HalfRuntimeConstFunc extends AbstractEqBinaryFunction {
+        private Utf8Sequence cachedRuntimeConst;
+        private long cachedSixPrefix;
 
-            return negated != Utf8s.equalsNc(a, b);
+        HalfRuntimeConstFunc(Function runtimeConst, Function arg) {
+            super(arg, runtimeConst);
+        }
+
+        @Override
+        public boolean getBool(Record rec) {
+            final Utf8Sequence a = left.getVarcharA(rec);
+            final Utf8Sequence b = cachedRuntimeConst;
+            if (a == null) {
+                return negated != (b == null);
+            }
+            return negated != Utf8s.equals(a, a.zeroPaddedSixPrefix(), b, cachedSixPrefix);
```

Comment:
At this point `b` can be `null` while `Utf8s.equals` has `@NotNull` annotation for the argument.